File: src/libs/m3/net/Socket.cc
/*
 * Copyright (C) 2021 Nils Asmussen, Barkhausen Institut
 * Copyright (C) 2021, Tendsin Mende <tendsin.mende@mailbox.tu-dresden.de>
 * Copyright (C) 2017, Georg Kotheimer <georg.kotheimer@mailbox.tu-dresden.de>
 *
 * This file is part of M3 (Microkernel-based SysteM for Heterogeneous Manycores).
 *
 * M3 is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as
 * published by the Free Software Foundation.
 *
 * M3 is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License version 2 for more details.
 */

#include <base/log/Lib.h>

#include <m3/net/Debug.h>
#include <m3/net/Net.h>
#include <m3/net/Socket.h>
#include <m3/net/TcpSocket.h>
#include <m3/net/UdpSocket.h>
#include <m3/session/NetworkManager.h>

#include <utility>

namespace m3 {

Socket::Socket(int sd, capsel_t caps, NetworkManager &nm)
    : File(0),
      _sd(sd),
      _state(Closed),
      _local_ep(),
      _remote_ep(),
      _nm(nm),
      _channel(caps),
      _recv_queue() {
}

Socket::~Socket() {
    _nm.abort(sd(), true);
}

void Socket::tear_down() noexcept {
    try {
        // we have no connection to tear down here, but only want to make sure that all packets we
        // sent are seen and handled by the server. thus, wait until we have got all replies to our
        // potentially in-flight packets, in which case we also have received our credits back.
        while(true) {
            fetch_replies();
            if(_channel.has_all_credits())
                break;

            log_net(NetLogEvent::StartedWaiting, _sd, 0);
            _channel.wait_for_credits();
            log_net(NetLogEvent::StoppedWaiting, _sd, 0);
        }
    }
    catch(...) {
        // ignore errors
    }
}

void Socket::disconnect() {
    _state = Closed;
    _local_ep = Endpoint();
    _remote_ep = Endpoint();
}

void Socket::process_message(const NetEventChannel::ControlMessage &message,
                             NetEventChannel::Event &event) {
    switch(message.type) {
        case NetEventChannel::Data:
            return handle_data(static_cast<NetEventChannel::DataMessage const &>(message), event);
        case NetEventChannel::Connected:
            return handle_connected(
                static_cast<NetEventChannel::ConnectedMessage const &>(message));
        case NetEventChannel::Closed:
            return handle_closed(static_cast<NetEventChannel::ClosedMessage const &>(message));
        case NetEventChannel::CloseReq:
            return handle_close_req(static_cast<NetEventChannel::CloseReqMessage const &>(message));
        default: throw Exception(Errors::NOT_SUP);
    }
}

void Socket::handle_data(NetEventChannel::DataMessage const &msg, NetEventChannel::Event &event) {
    log_net(NetLogEvent::RecvPacket, _sd, msg.size);
    LLOG(NET, "socket {}: received data with {}b from {}:{}"_cf, _sd, msg.size, IpAddr(msg.addr),
         msg.port);
    _recv_queue.append(new DataQueue::Item(&msg, std::move(event)));
}

void Socket::handle_connected(NetEventChannel::ConnectedMessage const &msg) {
    log_net(NetLogEvent::RecvConnected, _sd, msg.port);
    LLOG(NET, "socket {}: connected to {}:{}"_cf, _sd, IpAddr(msg.addr), msg.port);
    _state = Connected;
    _remote_ep.addr = IpAddr(msg.addr);
    _remote_ep.port = msg.port;
}

void Socket::handle_close_req(NetEventChannel::CloseReqMessage const &) {
    log_net(NetLogEvent::RecvRemoteClosed, _sd, 0);
    LLOG(NET, "socket {}: remote side was closed"_cf, _sd);
    _state = RemoteClosed;
}

void Socket::handle_closed(NetEventChannel::ClosedMessage const &) {
    log_net(NetLogEvent::RecvClosed, _sd, 0);
    LLOG(NET, "socket {}: closed"_cf, _sd);
    disconnect();
}

Option<std::tuple<const uchar *, size_t, Endpoint>> Socket::get_next_data() {
    while(true) {
        if(auto next = _recv_queue.get_next_data())
            return next;

        if(_state == Closed)
            throw Exception(Errors::INV_STATE);
        if(!_blocking) {
            process_events();
            return None;
        }

        wait_for_events();
    }
}

Option<std::pair<size_t, Endpoint>> Socket::do_recv(void *dst, size_t amount) {
    if(auto next = get_next_data()) {
        const auto [pkt_data, pkt_size, ep] = next.unwrap();
        size_t msg_size = Math::min(pkt_size, amount);
        memcpy(dst, pkt_data, msg_size);

        log_net(NetLogEvent::FetchData, _sd, msg_size);

        // ack read data and discard excess bytes that do not fit into the supplied buffer
        ack_data(msg_size);

        return Some(std::make_pair(msg_size, ep));
    }

    return None;
}

Option<size_t> Socket::do_send(const void *src, size_t amount, const Endpoint &ep) {
    // make sure that the message does not contain a page boundary
    ALIGNED(2048) char msg_buf[2048];
    Errors::Code res = _channel.build_data_message(msg_buf, sizeof(msg_buf), ep, src, amount);
    if(res != Errors::SUCCESS)
        throw Exception(res);

    while(true) {
        Errors::Code res = _channel.send_data(msg_buf, amount);
        if(res == Errors::SUCCESS) {
            log_net(NetLogEvent::SentPacket, _sd, amount);
            return Some(amount);
        }
        if(res != Errors::NO_CREDITS)
            throw Exception(res);

        if(!is_blocking()) {
            fetch_replies();
            return None;
        }

        wait_for_credits();

        if(_state == Closed)
            throw Exception(Errors::SOCKET_CLOSED);
    }
}

void Socket::ack_data(size_t size) {
    _recv_queue.ack_data(size);
}

void Socket::wait_for_events() {
    while(!process_events()) {
        log_net(NetLogEvent::StartedWaiting, _sd, 0);
        _channel.wait_for_events();
        log_net(NetLogEvent::StoppedWaiting, _sd, 0);
    }
}

void Socket::wait_for_credits() {
    while(true) {
        fetch_replies();
        if(can_send())
            break;

        log_net(NetLogEvent::StartedWaiting, _sd, 0);
        _channel.wait_for_credits();
        log_net(NetLogEvent::StoppedWaiting, _sd, 0);
    }
}

bool Socket::process_events() {
    bool seen_event = false;
    for(int i = 0; i < EVENT_FETCH_BATCH_SIZE; i++) {
        auto event = _channel.recv_message();
        if(!event.is_present())
            break;

        auto message = static_cast<NetEventChannel::ControlMessage const *>(event.get_message());
        process_message(*message, event);
        seen_event = true;
    }
    return seen_event;
}

void Socket::fetch_replies() {
    _channel.fetch_replies();
}

bool Socket::can_send() {
    return _channel.can_send();
}

}

File: src/server/net/src/smoltcpif/socket.rs
/*
 * Copyright (C) 2021 Nils Asmussen, Barkhausen Institut
 * Copyright (C) 2021, Tendsin Mende <tendsin.mende@mailbox.tu-dresden.de>
 *
 * This file is part of M3 (Microkernel-based SysteM for Heterogeneous Manycores).
 *
 * M3 is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as
 * published by the Free Software Foundation.
 *
 * M3 is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License version 2 for more details.
 */

use base::io::LogFlags;
use m3::cap::Selector;
use m3::errors::{Code, Error};
use m3::log;
use m3::mem::size_of;
use m3::net::{
    log_net, CloseReqMessage, ClosedMessage, ConnectedMessage, DataMessage, DataQueue, Endpoint,
    IpAddr, NetEvent, NetEventChannel, NetEventType, NetLogEvent, Port, Sd, SocketArgs, SocketType,
};
use m3::rc::Rc;
use m3::server::SessId;
use m3::time::{TimeDuration, TimeInstant};
use m3::vec;

use smoltcp::iface::SocketHandle;
use smoltcp::socket::{
    RawSocket, RawSocketBuffer, TcpSocket, TcpSocketBuffer, TcpState, UdpSocket, UdpSocketBuffer,
};
use smoltcp::storage::PacketMetadata;
use smoltcp::wire::IpVersion;
use smoltcp::wire::{IpAddress, IpEndpoint, Ipv4Address};

use crate::driver::DriverInterface;
use crate::ports::{AnyPort, EphemeralPort};

const CONNECT_TIMEOUT: TimeDuration = TimeDuration::from_secs(6);

pub fn to_m3_addr(addr: IpAddress) -> IpAddr {
    if addr.as_bytes().len() != 4 {
        IpAddr::unspecified()
    }
    else {
        let bytes = addr.as_bytes();
        IpAddr::new(bytes[0], bytes[1], bytes[2], bytes[3])
    }
}

/// Converts an IpEndpoint from smoltcp into an M³ (IpAddr, Port) tuple.
/// Assumes that the IpEndpoint a is Ipv4 address, otherwise this will panic.
pub fn to_m3_ep(addr: IpEndpoint) -> Endpoint {
    Endpoint::new(to_m3_addr(addr.addr), addr.port)
}

#[derive(Debug)]
pub enum SendNetEvent {
    Connected(ConnectedMessage),
    Closed(ClosedMessage),
    CloseReq(CloseReqMessage),
}

#[derive(Copy, Clone, Eq, PartialEq)]
pub enum State {
    Closed,
    Bound,
    Connecting,
    Connected,
    RemoteClosed,
}

/// Socket abstraction that unifies the different socket types
pub struct Socket {
    sd: Sd,
    socket: SocketHandle,
    ty: SocketType,
    state: State,
    connect_start: Option<TimeInstant>,
    _local_port: Option<EphemeralPort>,
    buffer_space: usize,

    // communication channel to client for incoming data/close-requests and outgoing events/data
    channel: Rc<NetEventChannel>,
    // pending incoming data events we could not send due to missing buffer space
    send_queue: DataQueue,
}

impl Socket {
    pub fn required_space(ty: SocketType, args: &SocketArgs) -> usize {
        args.rbuf_size
            + args.sbuf_size
            + match ty {
                SocketType::Dgram => {
                    (args.sbuf_slots + args.rbuf_slots) * size_of::<UdpSocketBuffer<'_>>()
                },
                SocketType::Raw => {
                    (args.sbuf_slots + args.rbuf_slots) * size_of::<RawSocketBuffer<'_>>()
                },
                _ => 0,
            }
    }

    pub fn new(
        sd: Sd,
        ty: SocketType,
        protocol: u8,
        args: &SocketArgs,
        caps: Selector,
        iface: &mut DriverInterface<'_>,
    ) -> Result<Self, Error> {
        let socket = match ty {
            SocketType::Stream => iface.add_socket(TcpSocket::new(
                TcpSocketBuffer::new(vec![0u8; args.rbuf_size]),
                TcpSocketBuffer::new(vec![0u8; args.sbuf_size]),
            )),
            SocketType::Dgram => iface.add_socket(UdpSocket::new(
                UdpSocketBuffer::new(vec![PacketMetadata::EMPTY; args.rbuf_slots], vec![
                    0u8;
                    args.rbuf_size
                ]),
                UdpSocketBuffer::new(vec![PacketMetadata::EMPTY; args.sbuf_slots], vec![
                    0u8;
                    args.sbuf_size
                ]),
            )),
            SocketType::Raw => iface.add_socket(RawSocket::new(
                IpVersion::Ipv4,
                protocol.into(),
                RawSocketBuffer::new(vec![PacketMetadata::EMPTY; args.rbuf_slots], vec![
                    0u8;
                    args.rbuf_size
                ]),
                RawSocketBuffer::new(vec![PacketMetadata::EMPTY; args.sbuf_slots], vec![
                    0u8;
                    args.sbuf_size
                ]),
            )),
            _ => return Err(Error::new(Code::InvArgs)),
        };

        Ok(Socket {
            sd,
            socket,
            ty,
            state: State::Closed,
            connect_start: None,
            _local_port: None,
            buffer_space: Self::required_space(ty, args),

            channel: NetEventChannel::new_server(caps)?,
            send_queue: DataQueue::default(),
        })
    }

    pub fn sd(&self) -> Sd {
        self.sd
    }

    pub fn socket_type(&self) -> SocketType {
        self.ty
    }

    pub fn channel(&self) -> &Rc<NetEventChannel> {
        &self.channel
    }

    pub fn buffer_space(&self) -> usize {
        self.buffer_space
    }

    pub fn fetch_event(&mut self, iface: &mut DriverInterface<'_>) -> Option<SendNetEvent> {
        match (self.ty, self.state) {
            (SocketType::Stream, State::Connecting) => {
                let tcp_socket = iface.get_socket::<TcpSocket<'_>>(self.socket);
                if tcp_socket.state() == TcpState::Established {
                    // disable Nagle's algorithm, because it delays sends, which at least for us
                    // reduces the achieved bandwidth in our benchmarks dramatically (factor 10).
                    // Maybe we don't transfer enough data?
                    tcp_socket.set_nagle_enabled(false);
                    if self.connect_start.take().is_some() {
                        crate::remove_timeout(self.socket);
                    }
                    self.state = State::Connected;
                    let ep = to_m3_ep(tcp_socket.remote_endpoint());
                    Some(SendNetEvent::Connected(ConnectedMessage::new(ep)))
                }
                // are we already trying to close the socket again?
                else if tcp_socket.state() != TcpState::Listen
                    && tcp_socket.state() != TcpState::SynSent
                    && tcp_socket.state() != TcpState::SynReceived
                {
                    self.state = State::RemoteClosed;
                    Some(SendNetEvent::CloseReq(CloseReqMessage::default()))
                }
                else if let Some(start) = self.connect_start {
                    if TimeInstant::now() >= start + CONNECT_TIMEOUT {
                        tcp_socket.abort();
                        self.connect_start = None;
                        crate::remove_timeout(self.socket);
                        self._local_port = None;
                        self.state = State::Closed;
                        self.send_queue.clear();
                        Some(SendNetEvent::Closed(ClosedMessage::default()))
                    }
                    else {
                        None
                    }
                }
                else {
                    None
                }
            },

            (SocketType::Stream, State::Connected | State::RemoteClosed) => {
                let tcp_socket = iface.get_socket::<TcpSocket<'_>>(self.socket);
                if !tcp_socket.is_open() {
                    self._local_port = None;
                    self.state = State::Closed;
                    self.send_queue.clear();
                    Some(SendNetEvent::Closed(ClosedMessage::default()))
                }
                // remote side has closed the connection?
                else if self.state != State::RemoteClosed
                    && tcp_socket.state() == TcpState::CloseWait
                    && !tcp_socket.can_recv()
                {
                    self.state = State::RemoteClosed;
                    Some(SendNetEvent::CloseReq(CloseReqMessage::default()))
                }
                else {
                    None
                }
            },

            _ => None,
        }
    }

    pub fn bind(
        &mut self,
        addr: IpAddress,
        port: AnyPort,
        iface: &mut DriverInterface<'_>,
    ) -> Result<(), Error> {
        if self.ty != SocketType::Dgram {
            return Err(Error::new(Code::InvArgs));
        }
        if self.state != State::Closed {
            return Err(Error::new(Code::InvState));
        }

        let endpoint = IpEndpoint::new(addr, port.number());
        let udp_socket = iface.get_socket::<UdpSocket<'_>>(self.socket);
        match udp_socket.bind(endpoint) {
            Ok(_) => {
                if let AnyPort::Ephemeral(e) = port {
                    self._local_port = Some(e);
                }
                self.state = State::Bound;
                Ok(())
            },
            Err(e) => {
                log!(LogFlags::Error, "bind failed: {}", e);
                // bind can only fail if the port is zero
                Err(Error::new(Code::InvArgs))
            },
        }
    }

    pub fn listen(
        &mut self,
        iface: &mut DriverInterface<'_>,
        addr: IpAddress,
        port: Port,
    ) -> Result<(), Error> {
        if self.ty != SocketType::Stream {
            return Err(Error::new(Code::InvArgs));
        }
        if self.state != State::Closed {
            return Err(Error::new(Code::InvState));
        }

        let endpoint = IpEndpoint::new(addr, port);
        let tcp_socket = iface.get_socket::<TcpSocket<'_>>(self.socket);
        match tcp_socket.listen(endpoint) {
            Ok(_) => {
                self.connect_start = None;
                self.state = State::Connecting;
                Ok(())
            },
            Err(e) => {
                log!(LogFlags::Error, "listen failed: {}", e);
                // listen can only fail if the port is zero
                Err(Error::new(Code::InvArgs))
            },
        }
    }

    pub fn connect(
        &mut self,
        remote_addr: IpAddr,
        remote_port: Port,
        local_port: EphemeralPort,
        iface: &mut DriverInterface<'_>,
    ) -> Result<(), Error> {
        if self.ty != SocketType::Stream {
            return Err(Error::new(Code::InvArgs));
        }
        if self.state != State::Closed {
            return Err(Error::new(Code::InvState));
        }

        let remote_endpoint = IpEndpoint::new(
            IpAddress::Ipv4(Ipv4Address::from_bytes(&remote_addr.0.to_be_bytes())),
            remote_port,
        );
        let local_endpoint = IpEndpoint::from(*local_port);

        let (tcp_socket, cx) = iface.get_socket_and_context::<TcpSocket<'_>>(self.socket);
        match tcp_socket.connect(cx, remote_endpoint, local_endpoint) {
            Ok(_) => {
                self.connect_start = Some(TimeInstant::now());
                crate::add_timeout(self.socket, TimeInstant::now() + CONNECT_TIMEOUT);
                self.state = State::Connecting;
                self._local_port = Some(local_port);
                Ok(())
            },
            Err(e) => {
                log!(LogFlags::Error, "connect failed: {}", e);
                // connect can only fail if the endpoints are invalid
                Err(Error::new(Code::InvArgs))
            },
        }
    }

    pub fn close(&mut self, iface: &mut DriverInterface<'_>) -> Result<(), Error> {
        if self.ty != SocketType::Stream {
            return Err(Error::new(Code::InvArgs));
        }

        let tcp_socket = iface.get_socket::<TcpSocket<'_>>(self.socket);
        tcp_socket.close();
        Ok(())
    }

    pub fn abort(&mut self, iface: &mut DriverInterface<'_>) {
        if self.ty == SocketType::Stream {
            let tcp_socket = iface.get_socket::<TcpSocket<'_>>(self.socket);
            tcp_socket.abort();
        }

        self._local_port = None;
        self.state = State::Closed;
    }

    pub fn receive<F>(&mut self, iface: &mut DriverInterface<'_>, func: F)
    where
        F: FnOnce(&[u8], IpEndpoint) -> usize,
    {
        match self.ty {
            SocketType::Stream => {
                let tcp_socket = iface.get_socket::<TcpSocket<'_>>(self.socket);
                if self.state == State::Connected || self.state == State::RemoteClosed {
                    let addr = tcp_socket.remote_endpoint();
                    // don't even log errors here, since they occur often and are uninteresting
                    tcp_socket
                        .recv(|d| {
                            if !d.is_empty() {
                                (func(d, addr), ())
                            }
                            else {
                                (0, ())
                            }
                        })
                        .ok();
                }
            },

            SocketType::Dgram => {
                let udp_socket = iface.get_socket::<UdpSocket<'_>>(self.socket);
                if let Ok((data, remote_endpoint)) = udp_socket.recv() {
                    func(data, remote_endpoint);
                }
            },

            SocketType::Raw => {
                let raw_socket = iface.get_socket::<RawSocket<'_>>(self.socket);
                if let Ok(data) = raw_socket.recv() {
                    func(data, IpEndpoint::UNSPECIFIED);
                }
            },

            SocketType::Undefined => panic!("cannot receive from undefined socket"),
        }
    }

    fn send(
        ty: SocketType,
        socket: SocketHandle,
        data: &[u8],
        dest_addr: IpAddr,
        dest_port: Port,
        iface: &mut DriverInterface<'_>,
    ) -> usize {
        match ty {
            SocketType::Stream => {
                let tcp_socket = iface.get_socket::<TcpSocket<'_>>(socket);
                if tcp_socket.can_send() {
                    tcp_socket.send_slice(data).unwrap()
                }
                else {
                    0
                }
            },

            SocketType::Dgram => {
                let udp_socket = iface.get_socket::<UdpSocket<'_>>(socket);
                if udp_socket.can_send() {
                    let rend = IpEndpoint::new(
                        IpAddress::Ipv4(Ipv4Address::from_bytes(&dest_addr.0.to_be_bytes())),
                        dest_port,
                    );

                    udp_socket.send_slice(data, rend).unwrap();
                    data.len()
                }
                else {
                    0
                }
            },

            SocketType::Raw => {
                let raw_socket = iface.get_socket::<RawSocket<'_>>(socket);
                if raw_socket.can_send() {
                    raw_socket.send_slice(data).unwrap();
                    data.len()
                }
                else {
                    0
                }
            },

            SocketType::Undefined => panic!("cannot send to undefined socket"),
        }
    }

    pub fn process_queued_events(&mut self, sess: SessId, iface: &mut DriverInterface<'_>) -> bool {
        let socket = self.socket;
        let ty = self.ty;
        let sd = self.sd;
        #[allow(clippy::blocks_in_conditions)]
        while self
            .send_queue
            .next_data(usize::MAX, &mut |data, ep: Endpoint| {
                let amount = Self::send(ty, socket, data, ep.addr, ep.port, iface);
                if amount > 0 {
                    log_net(NetLogEvent::SubmitData, sd, amount);
                    log!(
                        LogFlags::NetData,
                        "[{}] socket {}: sent delayed packet of {}b to {}",
                        sess,
                        sd,
                        amount,
                        ep,
                    );
                }
                (amount, amount)
            })
            .is_some()
        {}
        self.send_queue.has_data()
    }

    pub fn process_event(
        &mut self,
        sess: SessId,
        iface: &mut DriverInterface<'_>,
        event: NetEvent,
    ) -> bool {
        match event.msg_type() {
            NetEventType::Data => {
                let data = event.msg::<DataMessage>();
                let ip = IpAddr(data.addr as u32);
                let port = data.port as Port;

                let res = Self::send(
                    self.ty,
                    self.socket,
                    &data.data[0..data.size as usize],
                    ip,
                    port,
                    iface,
                );
                if res > 0 {
                    log_net(NetLogEvent::SubmitData, self.sd, res);
                    log!(
                        LogFlags::NetData,
                        "[{}] socket {}: sent packet of {}b to {}:{}",
                        sess,
                        self.sd,
                        res,
                        ip,
                        port,
                    );
                }

                if res < data.size as usize {
                    // if insufficient buffer space is available, remember the event for later
                    log!(
                        LogFlags::NetData,
                        "[{}] socket {}: no buffer space, delaying send of {}b to {}:{}",
                        sess,
                        self.sd,
                        data.size as usize - res,
                        ip,
                        port,
                    );
                    self.send_queue.append(event, res);
                    return true;
                }
            },

            NetEventType::CloseReq => {
                log!(
                    LogFlags::NetSess,
                    "[{}] net::close_req(sd={})",
                    sess,
                    self.sd
                );

                // ignore error
                self.close(iface).ok();
            },

            m => log!(LogFlags::Error, "Unexpected message from client: {:?}", m),
        }
        false
    }
}
